### ISSUE TITLE
关于inference的Decoding函数

### ISSUE THREAD

**Issue by KN1GHT9** [dhg-wei/DeCap#5]
在inference的Decoding函数中您的next_token是通过decoder解码出的logits计算得来，而后面进行文本解码的时候是使用了CLIP的tokenizer，但是这里是否存在一个问题：logits得到的张量有50257维，而CLIP的SimpleTokenizer仅能解码最大index为49407的token，会不会出现out-of-range的问题呢？我暂时的理解是你们使用了CLIP的token index来训练Decoder，在概率上避免了这个问题，但是没有在维度上面解决。刚进入该领域学习，如有理解错误请求指正。

**Comment by KN1GHT9** [comment]
~~~
def Decoding(model,clip_features):
    model.eval()
    embedding_cat = model.clip_project(clip_features).reshape(1,1,-1)
    entry_length = 30
    temperature = 1
    tokens = None
    for i in range(entry_length):
        outputs = model.decoder(inputs_embeds=embedding_cat)

        logits = outputs.logits
        logits = logits[:, -1, :] / (temperature if temperature > 0 else 1.0)
        logits_max = logits.max()
        logits = torch.nn.functional.softmax(logits)
        next_token = torch.argmax(logits, -1).unsqueeze(0)
        next_token_embed = model.decoder.transformer.wte(next_token)

        if tokens is None:
            tokens = next_token
        else:
            tokens = torch.cat((tokens, next_token), dim=1)
        if next_token.item()==49407:
            break
        embedding_cat = torch.cat((embedding_cat, next_token_embed), dim=1)
    try:
        output_list = list(tokens.squeeze().cpu().numpy())
        print(output_list)
        output = _Tokenizer.decode(output_list)
    except:
        output = 'None'
    return output
~~~

**Comment by dhg-wei** [comment]
You are right. As you mentioned, the logits of tokens that are large than 49407 will be very low. This avoids the out-of-range problem.
It's safer to make sure that the decoder's vocabulary length matches that of CLIP, which is 49408.
You can modify the decoder config file and retrain the model:
`config.vocab_size = 49408`
